Add peer visibility to MultiplayerSynchronizer.

Faless opened the pull request:
TL;DR;
---

Sets visibility for synchronized objects: `$MultiplayerSynchronizer.set_visibility_for(peer_id, visible)`

**or**

```gdscript
extends MultiplayerSynchronizer

func _enter_tree():
	add_visibility_filter(self._is_visible_to)

func _is_visible_to(peer) -> bool:
	return false # Hidden
```

The visibility is updated on every frame (default), physics frame, or manually.
You can further notify visibility changes by calling `$MultiplayerSynchronizer.update_visibility(peer:=0)` (where `0` means for all peers).

This also allows players to join mid-game (with some quirks).

Description
--

`MultiplayerSynchronizer`s can now be configured to limit their visibility to a subset of the connected peers, if the synchronized node was spawned by a `MultiplayerSpawner` (either automatically or via custom spawn) the given node will also be de-spawned remotely.

The replication system doesn't have the logic to handle sub-spawn directly, but it is possible to handle them appropriately by manually updating the visibility of the parent before changing the one of the nested spawns via the "update_visibility" function.

The visibility of each `MultiplayerSynchronizer` can specified by overriding `MultiplayerSynchronizer._is_visible_to(peer)`.

To further optimize the network code, visibility can be configured to be automatically updated during idle or physics, or set to always require manual update (`update_visibility` function).

See attached project:

![limited_visibility](https://user-images.githubusercontent.com/1687918/178558760-d369909e-056f-4e42-8e8c-49aaad8adb55.png)

~[proj_mp_interest.zip](https://github.com/godotengine/godot/files/9095533/proj_mp_interest.zip)~

~*Edit: updated* [proj_mp_interest2.zip](https://github.com/godotengine/godot/files/9106215/proj_mp_interest2.zip)~

*Edit: updated again* [proj_mp_interest3.zip](https://github.com/godotengine/godot/files/9113981/proj_mp_interest3.zip)


I also updated the tests in #62870 to work correctly with one notable hack (delaying world spawning players by one frame, so spawns get replicated in the correct order).

[test8.zip](https://github.com/godotengine/godot/files/9095584/test8.zip)

Note: this PR also adds basic support for **peers joining mid-game** (see demo above) although some more work is likely needed to improve cleanup, and potentially support single-player to multi-player transition.

Note for reviewers:

- MultiplayerSpawner `auto_spawn` property is removed, simply auto spawn whatever is in the "auto spawn list", if `_spawn_custom` is not overridden.
- I don't like having to change MultiplayerAPI, but that's needed for this PR. We should probably do more work and clean it up simplifying the interfaces and modularizing the default implementation as suggested in #62870, but I wanted to keep this PR light.
- There is still some work needed to cleanup the editor plugins, and make sure it shows a warning when the root paths are missing or unreachable for easier debugging.

Fixes https://github.com/godotengine/godot/issues/61711
Fixes https://github.com/godotengine/godot/issues/62320
Closes https://github.com/godotengine/godot-proposals/issues/3904

Faless committed:
Add peer visibility to MultiplayerSynchronizer.

MultiplayerSynchronizers can now be configured to limit their visibility
to a subset of the connected peers, if the synchronized node was spawned
by a MultiplayerSpawner (either automatically or via custom spawn) the
given node will also be despawned remotely.

The replication system doesn't have the logic to handle subspawn
directly, but it is possible to handle them appropriately by manually
updating the visibility of the parent before changing the one of the
nested spawns via the "update_visibility" function.

The visibility of each MultiplayerSynchronizer can be controlled by
adding or remove filters via "[add|remove]_visibility_filter(callable)".

To further optimize the network code, visibility filters can be configured
to be automatically updated during idle or physics frame, or set to always
require manual update (via the "update_visibility" function).

jgillich commented:
AFAIK the most common approach for this is going to be a proximity-based visibility implementation managed by the server. But since authority over MultiplayerSynchronizers is usually granted to the client, the server wouldn't be allowed to manage it, correct? 

The peer ID alone is rarely going to be enough to determine visibility so the `_is_visible_to` method has no use in this case, we would have to manage visibility using RPC instead.

reduz commented:
@jgillich AFAIK this logic only works server side, client has no authority. I am not entirely sure why you think the peer is not enough, you can obtain the node from it. Maybe one suggestion we could make to @Faless is to allow peers to register metadata (a Variant), so when you implement this function you can pass a pointer or something for other peers to get.

reduz commented:
@Faless After much though, I still think manually specifying visibility between pairs of peers instead of a callback is likely the way to go instead of this. Callback is OK for a simple game or for extra optional checks.

The reason is that if you have 100 or 1000 moving peers, you probably want to write your visibility code _outside_  your networking layer, then compute the changes between pairs and submit them to the networking layer. 

With the current approach, submitting the changes is quite inefficient because instead of doing.
```GDScript
peer_a.set_peer_visible(peer_b,false)
peer_b.set_peer_visible(peer_a,false)
```
you have to do

```GDScript
peer_a.update_visibility(peer_b)
peer_b.update_visibility(peer_b)
```
Then save that value _somewhere_, which will need some sort of indexing (a map), and read it back from your callback. Its really inefficient and cumbersome IMO and it will not scale as well.
```

jgillich commented:
> AFAIK this logic only works server side, client has no authority. I am not entirely sure why you think the peer is not enough, you can obtain the node from it.

Correct me if I'm wrong, but the client needs authority over its MultiplayerSynchronizers so it's allowed to push the values to the server. That's also how it's done in the [bomber example](https://github.com/godotengine/godot-demo-projects/blob/044afe7976735c628045759b3575b7175062535f/networking/multiplayer_bomber/player.gd#L21). Therefore, the server can't update visibility for clients. Would love to be wrong on this though :smile: (some docs on the new sync feature would be sweet, right now I'm mostly guessing based on examples and some of the C++ code that I don't fully understand)

reduz commented:
@jgillich In practice, Synchronizer is only for Server -> Peer, not the other way around. Ideally you want to do the client -> server communication via RPC so the server can validate what you are trying to do.

Faless commented:
I updated the PR with the simple visibility functions, and the callable as override.

The examples above becomes:

Much simpler for the simple case.

```gdscript
# Callback from Area node
func _player_visibility_enter(body):
    # Assuming body name is player id.
    set_peer_visibility(str(body.name).to_int(), true)

# Other area callback
func _player_visibility_exit(body):
    # Assuming body name is player id.
    set_peer_visibility(str(body.name).to_int(), false)
```

Pretty much the same for the complex scenario with a small change in callback registration.
```gdscript
# limited_sync.gd

func _enter_tree():
    # You can also set the override from a parent node.
    set_visibility_override(NetVisibility.is_visible_to.bind(self))
```

Faless commented:
> Correct me if I'm wrong, but the client needs authority over its MultiplayerSynchronizers so it's allowed to push the values to the server. That's also how it's done in the [bomber example](https://github.com/godotengine/godot-demo-projects/blob/044afe7976735c628045759b3575b7175062535f/networking/multiplayer_bomber/player.gd#L21).

A note on this, which I forgot to answer.
The example might indeed be misleading in this sense, but it was mostly shown as a possible configuration, not necessarily the best solution in most cases (and not even in that particular case to be fair). It just happens that the bomber demo has a fully disclosed network state and there is no need to limit peer visibility (i.e. everyone seeing everyone else is part of the game), so it was okay (and fast) to sync that way.

While you can still use replication to sync the "keyboard state" from the client to the server only (via the `set_peer_visibility(1, true)` function, or returning `true` for `1` in the callback), it's probably not worth the extra effort over RPCs, which also give you greater control.

I was also thinking about changing the bomber demo to make visibility matter, but it's also already quite overloaded of information, so we might need rethinking, or to be split it up as part of a tutorial of some kind.

Faless commented:
Okay, I've updated the PR so that callbacks behaves like filters (and updated the demos/snippet):

```gdscript
func _ready():
	$sync.add_visibility_filter(custom_visibility)

func custom_visibility(p_peer: int) -> bool:
	return true # Custom logic here

func block_filter(peer) -> bool: 
	return false # Always blocked

func _on_item_toggled(p_pressed):
	if p_pressed:
		# Block visibility from all
		$sync.add_visibility_filter(block_filter)
	else:
		# Remove visibility block
		$sync.remove_visibility_filter(block_filter)
```

akien-mga commented on core/multiplayer/multiplayer_api.h:
```diff
@@ -77,7 +77,8 @@ class MultiplayerCacheInterface : public RefCounted {
 
 	// Returns true if all peers have cached path.
-	virtual bool send_object_cache(Object *p_obj, NodePath p_path, int p_target, int &p_id) { return false; }
+	virtual bool send_object_cache(Object *p_obj, int p_target, int &p_id) { return false; }
```
```suggestion
	virtual bool send_object_cache(Object *p_obj, int p_target, int &r_id) { return false; }
```

akien-mga commented on core/multiplayer/multiplayer_api.h:
```diff
@@ -160,7 +161,8 @@ class MultiplayerAPI : public RefCounted {
 	Error replication_stop(Object *p_object, Variant p_config);
 	// Cache API
-	bool send_object_cache(Object *p_obj, NodePath p_path, int p_target, int &p_id);
+	bool send_object_cache(Object *p_obj, int p_target, int &p_id);
```
Same here.

akien-mga commented on doc/classes/MultiplayerSynchronizer.xml:
```diff
@@ -6,12 +6,64 @@
+		<method name="add_visibility_filter">
+			<return type="void" />
+			<argument index="0" name="filter" type="Callable" />
+			<description>
+			</description>
```
Would be great if you could write the docs too.

akien-mga replied:
As discussed this will be done in another PR.

akien-mga commented on scene/multiplayer/multiplayer_spawner.cpp:
```diff
@@ -71,7 +71,7 @@ bool MultiplayerSpawner::_get(const StringName &p_name, Variant &r_ret) const {
 
 void MultiplayerSpawner::_get_property_list(List<PropertyInfo> *p_list) const {
-	p_list->push_back(PropertyInfo(Variant::INT, "_spawnable_scene_count", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_EDITOR | PROPERTY_USAGE_ARRAY, "Scenes,scenes/"));
+	p_list->push_back(PropertyInfo(Variant::INT, "_spawnable_scene_count", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_EDITOR | PROPERTY_USAGE_ARRAY, "Auto Spawn List,scenes/"));
```
Curious to see if that breaks GDExtension too (https://github.com/godotengine/godot/issues/63066).
